pkg/solana/fees/block_history.go

```diff
@@ -133,5 +146,113 @@ func (bhe *blockHistoryEstimator) calculatePrice(ctx context.Context) error {
 		"slot", block.ParentSlot+1,
 		"count", len(feeData.Prices),
 	)
+
+	return nil
+}
+
+func (bhe *blockHistoryEstimator) calculatePriceFromMultipleBlocks(ctx context.Context, desiredBlockCount uint64) error {
+	// fetch client
+	c, err := bhe.client.Get()
+	if err != nil {
+		return fmt.Errorf("failed to get client: %w", err)
+	}
+
+	// Fetch the latest slot
+	currentSlot, err := c.SlotHeight(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get current slot: %w", err)
+	}
+
+	// Determine the starting slot for fetching blocks
+	if currentSlot < desiredBlockCount {
+		return fmt.Errorf("current slot is less than desired block count")
+	}
+	startSlot := currentSlot - desiredBlockCount + 1
+
+	// Fetch the last confirmed block slots
+	confirmedSlots, err := c.GetBlocksWithLimit(ctx, startSlot, desiredBlockCount)
+	if err != nil {
+		return fmt.Errorf("failed to get blocks with limit: %w", err)
+	}
+
+	// limit concurrency (avoid hitting rate limits)
+	semaphore := make(chan struct{}, 10)
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+	blockMedians := make([]ComputeUnitPrice, 0, desiredBlockCount)
+
+	// Iterate over the confirmed slots in reverse order to fetch most recent blocks first
+	// Iterate until we run out of slots
+	for i := len(*confirmedSlots) - 1; i >= 0; i-- {
+		slot := (*confirmedSlots)[i]
+
+		wg.Add(1)
+		go func(s uint64) {
+			defer wg.Done()
+			semaphore <- struct{}{}
+			defer func() { <-semaphore }()
+
+			// Fetch the block details
+			block, errGetBlock := c.GetBlock(ctx, s)
+			if errGetBlock != nil || block == nil {
+				// Failed to get block at slot || no block found at slot
+				if errGetBlock != nil {
+					bhe.lgr.Errorw("BlockHistoryEstimator: failed to get block", "slot", s, "error", errGetBlock)
+				}
+				return
+			}
```

Comment:
nit: maybe we can split this into 2 different if's to avoid extra nesting. We can also log an error for the other scenario too
```suggestion
			if errGetBlock != nil {
				// Failed to get block at slot
			        bhe.lgr.Errorw("BlockHistoryEstimator: failed to get block", "slot", s, "error", errGetBlock)
				return
			}
			if block == nil {
			  // No block found at slot
			  bhe.lgr.Errorw("BlockHistoryEstimator: failed to get block at slot", "slot", s)
			  return
			}
```

Reply:
Thought the `block == nil` case would be too noisy as not every slot may have a block. Just tested with my custom script across 100 txes and didn't return any log so I'll include it

Reply:
Actually that's a fair point. You can skip the log for `block == nil` but we should still split it into different if's